Add native job support for neovim.

okkays opened the pull request:
Hello!

I think I've managed to get support for neovim's jobcontrol (#125) working (I've tested it manually with [vim-coverage](https://github.com/google/vim-coverage), which wasn't working for me and caused me to try and solve this in the first place).

However, in order to get system-vimjob.vroom to succeed, I need to pass vroom `--shell-delay .25` (by default vim's delay is 0.25 and neovim's is 0.0).  Here's an example failure:

```
└─ $ vroom --neovim vroom/system-vimjob.vroom
vroom/system-vimjob.vroom
FAILED on line 45: Multiple failures:
Expected system call not received.

Got no chance to inject response:
 EXPECT:        echo hi (regex mode)

Failed command on line 45:
:let g:invocation = AsyncWait(g:syscall.CallAsync('Callback', 0), 2)

Queued system controls are:
 EXPECT:        echo hi (regex mode)

No system calls received. Perhaps your --shell is broken?

Last few commands (most recent last) were:
:if has('nvim')<CR>  set cmdheight=30<CR>endif<CR><CR>
:set nocompatible<CR>
:let g:maktabadir = fnamemodify($VROOMFILE, ':p:h:h')<CR>
:let g:bootstrapfile = g:maktabadir . '/bootstrap.vim'<CR>
:execute 'source' g:bootstrapfile<CR>
:call maktaba#system#SetUsableShellRegex('\v<shell\.vroomfaker$')<CR>
:if !has('job') && !has('nvim')<CR>  echomsg maktaba#error#MissingFeature(      'Must have +job support to run system-vimjob.vroom examples')<CR>endif<CR>
:function AsyncWait(invocation, delay) abort<CR>  let l:deadline = localtime() + a:delay " wait for at most N seconds<CR>  while !a:invocation.finished && localtime() < l:deadline<CR>    sleep 100m<CR>  endwhile<CR>  call maktaba#ensure#IsTrue(a:invocation.finished, 'Async callback was expected to complete within %d seconds', a:delay)<CR>  return a:invocation<CR>endfunction<CR>
:let g:syscall = maktaba#syscall#Create(['echo', 'hi'])<CR>
:function Callback(result) abort<CR>  let g:callback_stdout = a:result.stdout<CR>endfunction<CR>
:let g:invocation = AsyncWait(g:syscall.CallAsync('Callback', 0), 2)<CR>

Ran 1 test in vroom/system-vimjob.vroom. 0 passing, 0 errored, 1 failed.
```

Again, this works fine:

```
└─ $ vroom --neovim vroom/system-vimjob.vroom --shell-delay .25
vroom/system-vimjob.vroom
Ran 1 test in vroom/system-vimjob.vroom. 1 passing, 0 errored, 0 failed.
```

Any ideas on whether this is an actual problem in my use of `jobcontrol`, or if the async delay is just necessary for nvim in this case?  I've looked into it for a bit, but this is my first adventure in maktaba/vroom.

okkays commented:
@googlebot I signed it!

dbarnett commented:
Sure, I can look into why.

For starters you might hack the AsyncAwait helper in the test to change the loop from
```
  |  while !a:invocation.finished && localtime() < l:deadline<CR>
  |    sleep 100m<CR>
  |  endwhile<CR>
```
to 
```
  |  let l:finished_log = []<CR>
  |  while !a:invocation.finished && localtime() < l:deadline<CR>
  |    call add(l:finished_log, a:invocation.finished)<CR>
  |    sleep 100m<CR>
  |  endwhile<CR>
  |  echomsg l:finished_log
```
and see what that spits out.

It would be strange if at_exit was getting called before the syscall process actually finished writing its files... maybe a slight bug with neovim jobcontrol?

okkays commented:
finished_log contains a single 0 at that point - it looks like it's finishing within 100ms (though not before getting to that loop).

The jobcontrol docs actually [explicitly mention](https://neovim.io/doc/user/job_control.html#on_exit) this kind of bug when using callbacks, but suggests [channel-buffered](https://neovim.io/doc/user/channel.html#channel-buffered) as a workaround (which I'm using).

If I look at `self.stdout` in `maktaba#syscall#neovim#HandleJobExit`, it actually contains `hi` - so it looks like nvim is doing what it's supposed to and populating stdout on time.  That would make me think that `jobstart` isn't using `shell.vroomfaker`, but since it works fine with the delay, maybe it's a bug in how we're using vroom to capture neovim syscalls?  Something there would also explain why it works fine in practice, too.

dbarnett commented:
K, have you tried adding a delay control in the vroom file to just the one line to see if that takes care of failures? As long as we have the tests passing consistently and a note explaining what's going on, I'm fine with proceeding now that you've done some initial investigation.

Do consider either renaming system-vimjob.vroom or splitting into vim/neovim variants. The "vimjob" was referring specifically to `+vimjob`.

okkays commented:
Went ahead and added some delays.  I'm really liking vroom's style and syntax, nice job on that :+1:

I've renamed the test to `system-job.vroom` (`:help job` will take users to the right place in both vim and nvim).

I chose to rename instead of split because of a note in the issue (#125):

> Note this might complicate testing since the actual system calls would be different between vim and neovim, but a single vroom file would have to test them both.

which (barring the delays) makes sense to me, because both applications should have the same API, and this way there won't need to be special handling to skip each split file for one or the other (at least, whenever the tests are updated to use vim8). Plus, changes to that API only need to be tested in one place.

lmk if there's anything else to tweak

dbarnett commented on vroom/system-job.vroom:
```diff
@@ -35,9 +47,9 @@ To execute system calls asynchronously, use CallAsync.
+  :let g:invocation = AsyncWait(g:syscall.CallAsync('Callback', 0), 2) (0.1s)
   ! echo hi
-  :echomsg g:callback_stdout
+  :echomsg g:callback_stdout (0.1s)
```
Strange that it's needed on these `:echomsg` lines. What's the effect of not having it? Does making the previous time longer and dropping this one work?

okkays replied:
Yep, good call.  They were evidently just acting to extend the CallAsync delays, since 100ms wasn't enough.

dbarnett commented:
> Went ahead and added some delays. I'm really liking vroom's style and syntax, nice job on that 👍

That's credit to @Soares who pushed for literate testing and built the initial implementation. =)
